src/charm.py

```diff
@@ -314,6 +321,46 @@ def _reconcile(self):
         remote_write.reload_alerts()
         self._add_remote_write(remote_write.endpoints, insecure_skip_verify)
 
+        # Enable traces ingestion with TracingEndpointProvider, i.e. configure the receivers
+        tracing_provider = TracingEndpointProvider(self, relation_name="receive-traces")
+        requested_tracing_protocols = set(tracing_provider.requested_protocols()).union(
+            {
+                receiver
+                for receiver in get_args(ReceiverProtocol)
+                if self.config.get(f"always_enable_{receiver}")
+            }
+        )
+        # Send tracing receivers over relation data to charms sending traces to otel collector
+        if self.unit.is_leader():
+            tracing_provider.publish_receivers(
+                tuple(
+                    (
+                        protocol,
+                        self._get_tracing_receiver_url(
+                            protocol=protocol,
+                            tls_enabled=is_server_cert_on_disk(container),
+                        ),
+                    )
+                    for protocol in requested_tracing_protocols
+                )
+            )
+        self._add_traces_ingestion(requested_tracing_protocols)
+        # Add default processors to traces
+        self._add_traces_processing()
+        # Enable pushing traces to a backend (i.e. Tempo) with TracingEndpointRequirer, i.e. configure the exporters
+        tracing_requirer = TracingEndpointRequirer(
+            self,
+            relation_name="send-traces",
+            protocols=[
+                "otlp_http",  # for charm traces
+                "otlp_grpc",  # for forwarding workload traces
+            ],
+        )
```

Comment:
The above seems well contained. How about
```python
tracing_provider = enable_tracing_backend()
```

Having `_add_traces_ingestion` taking `self` seems like our old ways.

Reply:
I'm confused; passing `self` is our "new ways" as well, all the other integrations happen that way:
- `self._add_self_scrape()`
- `self._add_remote_write()`
- `self._add_log_ingestion()`
- `self._add_log_forwarding()`
- `self._add_cloud_integrator()`

If you mean the tracing provider itself, that's also in line with the other relation objects, because the charm libraries are still the same
- `LokiPushApiConsumer(self, ...)`
- `LokiPushApiProvider(self, ...)`
- `MetricsEndpointConsumer(self)`
- `PrometheusRemoteWriteConsumer(self, ...)`
- `GrafanaCloudConfigRequirer(self, ...)`

Our "old ways" was **assigning those objects to self**, which we don't to for tracing either